consensus/src/marshal/mod.rs

```diff
@@ -0,0 +1,336 @@
+        runner.start(|mut context| async move {
+            // Initialize network
+            let (network, mut oracle) = Network::new(
+                context.with_label("network"),
+                simulated::Config {
+                    max_size: 1024 * 1024,
+                },
+            );
+            network.start();
+
+            // Generate private keys and sort them by public key.
+            let mut schemes = (0..NUM_VALIDATORS)
+                .map(|i| PrivateKey::from_seed(i as u64))
+                .collect::<Vec<_>>();
+            schemes.sort_by_key(|s| s.public_key());
+            let peers: Vec<PublicKey> = schemes.iter().map(|s| s.public_key()).collect();
+
+            // Generate shares
+            let (identity, shares) =
+                generate_shares::<_, V>(&mut context, None, NUM_VALIDATORS, QUORUM);
+            let identity = *poly::public::<V>(&identity);
+
+            // Initialize validators
+            let mut pks = Vec::new();
+            let mut secrets = Vec::new();
+            for scheme in schemes.iter() {
+                pks.push(scheme.public_key());
+                secrets.push(scheme);
+            }
+
+            // Initialize applications and actors
+            let mut applications = BTreeMap::new();
+            let mut actors = Vec::new();
+            let coordinator = resolver::mocks::Coordinator::new(pks.clone());
+
+            for i in 0..NUM_VALIDATORS as usize {
+                let (application, actor) = setup_validator(
+                    context.with_label(&format!("validator-{i}")),
+                    &mut oracle,
+                    coordinator.clone(),
+                    secrets[i].clone(),
+                    identity,
+                )
+                .await;
+                applications.insert(pks[i].clone(), application);
+                actors.push(actor);
+            }
+
+            // Add links between all peers
+            let link = Link {
+                latency: NETWORK_SPEED.as_millis() as f64,
+                jitter: JITTER,
+                success_rate: SUCCESS_RATE,
+            };
+            for p1 in peers.iter() {
+                for p2 in peers.iter() {
+                    if p2 == p1 {
+                        continue;
+                    }
+                    oracle
+                        .add_link(p1.clone(), p2.clone(), link.clone())
+                        .await
+                        .unwrap();
+                }
+            }
+
+            // Generate blocks, skipping the genesis block.
+            let mut blocks = Vec::<B>::new();
+            let mut parent = sha256::hash(b"");
+            for i in 1..=NUM_BLOCKS {
+                let block = B::new::<sha256::Sha256>(parent, i, i);
+                parent = block.digest();
+                blocks.push(block);
+            }
+
+            // Broadcast and finalize blocks in random order
+            blocks.shuffle(&mut context);
```

Comment:
nit: we could have this test take a `seed` argument and the test itself could run it with a bunch of different seeds?

Like: https://github.com/commonwarexyz/monorepo/blob/868a146588cca5ecc164c9f935f68d7ec0d67ef4/consensus/src/threshold_simplex/mod.rs#L1907

This would also allow us to add a simple determinism test (think of this as required IMO).